tests/filecheck/dialects/csl/csl-stencil.md

```diff
@@ -0,0 +1,303 @@
+//                         at [offset, offset+chunkSize] in the sender's send_dsd.
+//                         to access data in the receive buffers, use getRecvBufDsd(), getRecvBufDsdByDir(), or getRecvBufDsdByNeighbor().
+//                         the total number of invocations of the callback equals numChunks amount of time.
+//   - communicate_cb:     callback invoked once after all communication (sending and receiving) is completed.
+
+fn communicate(send_dsd: mem1d_dsd,
+               num_chunks: i16,
+               clear_recv_buf_cb: *const fn(i16)void,
+               communicate_cb: *const fn()void
+              ) void { ... }
+```
+
+## Step 2: Communication with compute
+
+The above stencil comms CSL function is invoked with two callbacks with distinct tasks, and
+it's best to understand them from their types:
+* The `clear_recv_buf_cb` takes a collection of partial z-value tensors and reduces them to one
+  partial z-value tensor. It is invoked one for each communicated slice (essentially acting like a loop body),
+  and builds up one combined full (non-partial) z-value tensor.
+  * Note: Both `csl_stencil.prefetch` and `csl_stencil.apply` are lowered to invoke the same (above)
+    function with `clear_recv_buf_cb` callback - the difference is that prefetch copies the data
+    received from `n` neighbours into `n` buffers, while `csl_stencil.apply` consumes the data
+    received from `n` neighbours into 1 buffer by applying the compute given in the `stencil.apply` body.
+* The `communicate_cb` takes the reduced data and/or that of any prefetched buffers, applying
+  any computation that can only be done after communication has finished. It also performs
+  the semantic task of `stencil.return`.
+
+The following shows `csl_stencil.apply`, combining the functionality of `csl_stencil.prefetch`
+and `stencil.apply`, with two code blocks corresponding to the two callbacks (in order):
+
+```
+builtin.module {
+  func.func @gauss_seidel_func(%a : !stencil.field<[-1,1023]x[-1,511]xtensor<512xf32>>, %b : !stencil.field<[-1,1023]x[-1,511]xtensor<512xf32>>) {
+    %0 = stencil.load %a : !stencil.field<[-1,1023]x[-1,511]xtensor<512xf32>> -> !stencil.temp<[-1,2]x[-1,2]xtensor<512xf32>>
+
+    %1 = tensor.empty() : tensor<510xf32>
+    %2 = csl_stencil.apply(%3 = %0 : !stencil.temp<[-1,2]x[-1,2]xtensor<512xf32>>, %empty_res = %1 : tensor<510xf32>) {"topo" = #dmp.topo<1022x510>, "size" = 510, "swaps" = [
+            #csl_stencil.exchange<to [1, 0]>,
+            #csl_stencil.exchange<to [-1, 0]>,
+            #csl_stencil.exchange<to [0, 1]>,
+            #csl_stencil.exchange<to [0, -1]>
+        ]} -> (!stencil.temp<[0,1]x[0,1]xtensor<510xf32>>) ({
+      ^0(%recv : !stencil.temp<[-1,2]x[-1,2]xtensor<255xf32>>, %offset : i64, %iter_arg : tensor<510xf32>):
+        // reduces chunks from neighbours into one chunk (clear_recv_buf_cb)
+        %3 = csl_stencil.access %recv[1, 0] : !stencil.temp<[-1,2]x[-1,2]xtensor<255xf32>>
```

Comment:
Also, here you re-define `%3` which makes it less readable again^^

Reply:
Updated